overseer/src/lib.rs

```diff
@@ -0,0 +1,671 @@
+	/// receiver. The most obvious use case of this is communicating with children.
+	SubsystemMessage {
+		to: Option<I>,
+		msg: M,
+	},
+	/// A message that wraps something the `Subsystem` is desiring to
+	/// spawn on the overseer and a `oneshot::Sender` to signal the result
+	/// of the spawn.
+	SpawnJob {
+		s: BoxFuture<'static, ()>,
+		res: oneshot::Sender<SubsystemResult<()>>,
+	},
+}
+
+enum Event {
+	BlockImport,
+	BlockFinalized,
+	Stop,
+}
+
+/// A handler used to communicate with the `Overseer`.
+#[derive(Clone)]
+pub struct OverseerHandler {
+	events_tx: mpsc::Sender<Event>,
+}
+
+impl OverseerHandler {
+	/// Inform the `Overseer` that that some block was imported.
+	pub async fn block_imported(&mut self) -> SubsystemResult<()> {
+		self.events_tx.send(Event::BlockImport).await.map_err(|_| SubsystemError)
+	}
+
+	/// Inform the `Overseer` that that some block was finalized.
+	pub async fn block_finalized(&mut self) -> SubsystemResult<()> {
+		self.events_tx.send(Event::BlockFinalized).await.map_err(|_| SubsystemError)
+	}
+
+	/// Tell `Overseer` to shutdown.
+	pub async fn stop(&mut self) -> SubsystemResult<()> {
+		self.events_tx.send(Event::Stop).await.map_err(|_| SubsystemError)
+	}
+}
+
+impl<M: Debug, I: Debug> Debug for OverseerMessage<M, I> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		match self {
+			OverseerMessage::SubsystemMessage { to, msg } => {
+				write!(f, "OverseerMessage::SubsystemMessage{{ to: {:?}, msg: {:?} }}", to, msg)
+			}
+			OverseerMessage::SpawnJob { .. } => write!(f, "OverseerMessage::Spawn(..)")
+		}
+	}
+}
+
+/// A running instance of some `Subsystem`.
+struct SubsystemInstance<M: Debug, I> {
+	/// We talk to the `Overseer` over this channel.
+	rx: mpsc::Receiver<OverseerMessage<M, I>>,
+	/// The `Overseer` talks to use over this channel.
+	tx: mpsc::Sender<M>,
+}
+
+/// A context type that is given to the `Subsystem` upon spawning
+/// that can be used by `Subsystem` to communicate with the outside world.
+pub struct SubsystemContext<M: Debug, I>{
+	rx: mpsc::Receiver<M>,
+	tx: mpsc::Sender<OverseerMessage<M, I>>,
+}
+
+impl<M: Debug, I> SubsystemContext<M, I> {
+	/// Try to asyncronously receive a message.
+	///
+	/// This has to be used with caution, if you loop over this without
+	/// using `pending!()` macro you will end up with a busy loop!
+	pub async fn try_recv(&mut self) -> Result<Option<M>, ()> {
+		match poll!(self.rx.next()) {
+			Poll::Ready(Some(msg)) => Ok(Some(msg)),
+			Poll::Ready(None) => Err(()),
+			Poll::Pending => Ok(None),
+		}
+	}
+
+	/// Receive a message.
+	pub async fn recv(&mut self) -> SubsystemResult<M> {
+		self.rx.next().await.ok_or(SubsystemError)
+	}
+
+	/// Send a message to whom it may concern.
+	///
+	/// The message will be broadcasted to all other `Subsystem`s that can
+	/// receive it.
+	pub async fn broadcast_msg(&mut self, msg: M) {
+		let _ = self.tx.send(OverseerMessage::SubsystemMessage{
```

Comment:
The guide doesn't speak about broadcast messaging but only about targeted messaging. This is one of the reasons why I've advocated for a "big struct" approach as opposed to type-erasure from the beginning, as it makes it easier for the overseer to determine which messages go to which subsystems.

The main difficulty I foresee with broadcast is that it makes responses harder to program in. I was assuming we might have messages like this:
```rust
enum PoVFetchMessage {
    FetchPoVBlock(CandidateReceipt, oneshot::Sender<PoVBlock>)
}
```

In the broadcast case, you are requiring `clone`, which `oneshot::Sender` does not provide. How should we work around that?

Reply:
The one alternative that I see is to bake in a `from` parameter, so we'd have

```rust
enum PoVFetchMessage {
    FetchPoVBlock(CandidateReceipt, SubsystemAddress),
}
```

and the `PoVFetch` can use the address to determine who to send the reply to. However, this makes communication with subsystems from outside difficult.

One concrete use-case we have is for a block-authorship-data-provider subsystem to give data to the BABE block authorship `Proposer` as necessary.

```rust
// in proposer
let (tx, rx) = oneshot::channel();
overseer.dispatch_message(
    BlockAuthorshipMessage::GetBlockAuthorshipData(parent_hash).into(),
    tx,
).await;

let (bitfields, backed_candidates) = rx.await.ok().unwrap_or_default();
```

Reply:
`from` can be `Option`